webpack/assets/javascripts/react_app/components/MemoryAllocationInput/MemoryAllocationInput.js

```diff
@@ -0,0 +1,143 @@
+import React, { useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+import { WarningTriangleIcon, ErrorCircleOIcon } from '@patternfly/react-icons';
+import { HelpBlock, Grid, Col, Row } from 'patternfly-react';
+import { translate as __ } from '../../common/I18n';
+import NumericInput from '../common/forms/NumericInput';
+import { GB_FORMAT, MB_FORMAT, GB_STEP, MB_STEP } from './constants';
+import './memoryAllocationInput.scss';
+import { noop } from '../../common/helpers';
+
+const MemoryAllocationInput = ({
+  label,
+  defaultValue,
+  onChange,
+  maxValue,
+  recommendedMaxValue,
+}) => {
+  const [value, setValue] = useState(defaultValue);
+  const [validationState, setValidationState] = useState(null);
+
+  useEffect(() => {
+    if (value > maxValue && maxValue != null) {
+      setValidationState('error');
+    } else if (value > recommendedMaxValue && recommendedMaxValue != null) {
+      setValidationState('warning');
+    } else {
+      setValidationState(null);
+    }
+  }, [validationState, recommendedMaxValue, maxValue, value]);
+
+  const handleValueIncrease = () => {
+    if (value >= GB_STEP) {
+      setValue(value + GB_STEP);
+    } else {
+      setValue(value + MB_STEP);
+    }
+  };
+
+  const handleValueDecrease = () => {
+    if (value <= GB_STEP) {
+      setValue(value - MB_STEP);
+    } else {
+      setValue(value - GB_STEP);
+    }
+  };
+
+  const handleTypedValue = v => {
+    if (v > GB_STEP) {
+      setValue(Math.round(v / GB_STEP) * GB_STEP);
+    } else {
+      setValue(Math.round(v / MB_STEP) * MB_STEP);
+    }
+  };
+
+  const handleChange = v => {
+    if (v === value + 1) {
+      handleValueIncrease();
+    } else if (v === value - 1) {
+      handleValueDecrease();
+    } else {
+      handleTypedValue(v);
+    }
+    onChange(value);
+  };
+
+  const format = v => {
+    // used the buttons
+    if (v % MB_STEP === 0) {
+      if (v >= GB_STEP) {
+        return `${v / GB_STEP} ${GB_FORMAT}`;
+      }
+      return `${v} ${MB_FORMAT}`;
+    }
+
+    // typed value
+    return v;
+  };
+
+  const helpBlock = () => {
+    if (validationState === 'warning') {
+      return (
+        <HelpBlock>
+          <WarningTriangleIcon className="warning-icon" />
+          {__('Specified value is higher than recommended maximum')}
+        </HelpBlock>
+      );
+    } else if (validationState === 'error') {
+      return (
+        <HelpBlock>
+          <ErrorCircleOIcon className="error-icon" />
+          {__('Specified value is higher than maximum value')}
+        </HelpBlock>
+      );
+    }
+    return null;
+  };
+
```

Comment:
Can this be tested with snapshots?

Reply:
I added tests to validate the warning and error alert.